contracts/true-currencies/TrueCurrencyWithPoR.sol

```diff
@@ -0,0 +1,98 @@
+            super._mint(account, amount);
+            return;
+        }
+
+        // Get latest proof-of-reserves from the feed
+        (, int256 answer, , uint256 updatedAt, ) = IChainlinkAggregatorV3(chainReserveFeed).latestRoundData();
+        require(answer > 0, "TrueCurrency: Invalid answer from PoR feed");
+
+        // Check the answer is fresh enough (i.e., within the specified heartbeat)
+        uint256 oldestAllowed = block.timestamp.sub(chainReserveHeartbeat, "TrueCurrency: Invalid timestamp from PoR feed");
+        require(updatedAt >= oldestAllowed, "TrueCurrency: PoR answer too old");
+
+        // Get required info about total supply & decimals
+        uint256 currentSupply = totalSupply();
+        uint8 trueDecimals = decimals();
+        uint8 reserveDecimals = IChainlinkAggregatorV3(chainReserveFeed).decimals();
+        uint256 reserves = uint256(answer);
+        // Normalise TrueCurrency & reserve decimals
+        if (trueDecimals < reserveDecimals) {
+            currentSupply = currentSupply.mul(10**uint256(reserveDecimals - trueDecimals));
+        } else if (trueDecimals > reserveDecimals) {
+            reserves = reserves.mul(10**uint256(trueDecimals - reserveDecimals));
+        }
+
+        // Check that after minting more tokens, the total supply would NOT exceed the reserves
+        // reported by the latest valid proof-of-reserves feed.
+        require(currentSupply + amount <= reserves, "TrueCurrency: total supply would exceed reserves after mint");
+        super._mint(account, amount);
+    }
+
+    /**
+     * @notice Sets a new feed address
+     * @dev Admin function to set a new feed
+     * @param newFeed Address of the new feed
+     */
+    function setChainReserveFeed(address newFeed) external override onlyOwner returns (uint256) {
+        require(newFeed != chainReserveFeed, "TrueCurrency: new chainReserveFeed must be different to current feed");
+
+        emit NewChainReserveFeed(chainReserveFeed, newFeed);
+        chainReserveFeed = newFeed;
+    }
+
+    /**
+     * @notice Sets the feed's heartbeat expectation
+     * @dev Admin function to set the heartbeat
+     * @param newHeartbeat Value of the age of the latest update from the feed
+     */
+    function setChainReserveHeartbeat(uint256 newHeartbeat) external override onlyOwner returns (uint256) {
+        require(newHeartbeat <= MAX_AGE, "TrueCurrency: PoR heartbeat greater than MAX_AGE");
+
+        // Allowable scenarios:
+        //  - heartbeat is not initialised (0 instead of default MAX_AGE); OR
+        //  - new heartbeat is different AND
+        //    new heartbeat is not resetting to default while current heartbeat is already set to the default
+        require(
+            chainReserveHeartbeat == 0 ||
+                (newHeartbeat != chainReserveHeartbeat && !(newHeartbeat == 0 && chainReserveHeartbeat == MAX_AGE)),
+            "TrueCurrency: new chainReserveHeartbeat must be different to current heartbeat"
```

Comment:
Add test cases covering this condition.

Reply:
Do we need this require? It shouldn't hurt to set chainReserveHeartbeat to the same value as before, and if we want an explicit `MIN_AGE == 1` then we should add that as a constant -- although personally I don't see much difference between 1 second of heartbeat and 0 seconds for Chainlink's update frequency (at most once per block).

Reply:
Added tests covering the condition.

@yuchenlintt That's a good question. The code was originally written by a Chainlink eng. It seems for some reason he thought that not allowing to set `chainReserveHeartbeat` to the same value was important. I can see this would lead to emission of `NewChainReserveHeartbeat(<n>, <n>)` events, which would be a little silly, but we could as well skip emitting instead of reverting in this case. What do you think?

Reply:
I think it's fine to emit an event that says nothing changed. Anyone who consumes the event should be able to handle this without any issues.

In smart contracts I'd much prefer to reason about simple code than a bunch of edge cases. And in this example I don't see any failure modes that would force us to add this require (which might instead backfire and prevent us from newHeartbeat == 0 if for some reason we need to, for instance).

Reply:
Committed suggestion.